server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -0,0 +1,257 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.health.node.selection.HealthNode;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class periodically logs the results of the Health API to the standard Elasticsearch server log file.
+ */
+public class HealthPeriodicLogger implements ClusterStateListener, Closeable, SchedulerEngine.Listener {
+    public static final String HEALTH_FIELD_PREFIX = "elasticsearch.health";
+
+    /**
+     * Creates a new HealthPeriodicLoggerResult.
+     * This creates a record that's used as the output of this feature.
+     *
+     * @param indicatorResults the results of the Health API call that will be used as the output.
+     */
+    public record HealthPeriodicLoggerResult(List<HealthIndicatorResult> indicatorResults) {
+
+        /**
+         * Create a Map of the results, which is then turned into JSON for logging
+         */
+        public Map<String, Object> toMap() {
+            final Map<String, Object> result = new HashMap<>();
+
+            if (this.indicatorResults != null && this.indicatorResults.size() > 0) {
+                // overall status
+                final HealthStatus status = HealthStatus.merge(this.indicatorResults.stream().map(HealthIndicatorResult::status));
+                result.put("elasticsearch.health.overall.status", status.xContentValue());
+
+                // top-level status for each indicator
+                this.indicatorResults.forEach((indicatorResult) -> {
+                    result.put(
+                        String.format(Locale.ROOT, "%s.%s.status", HEALTH_FIELD_PREFIX, indicatorResult.name()),
+                        indicatorResult.status().xContentValue()
+                    );
+                });
+            }
+
+            return result;
+        }
+    }
+
+    public static final String HEALTH_PERIODIC_LOGGER_POLL_INTERVAL = "health_periodic_logger.poll_interval";
+    public static final Setting<TimeValue> HEALTH_PERIODIC_LOGGER_POLL_INTERVAL_SETTING = Setting.timeSetting(
+        HEALTH_PERIODIC_LOGGER_POLL_INTERVAL,
+        TimeValue.timeValueSeconds(60),
+        TimeValue.timeValueSeconds(15),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Name constant for the job HealthService schedules
+     */
+    protected static final String HEALTH_PERIODIC_LOGGER_JOB_NAME = "health_periodic_logger";
+
+    private final Settings settings;
+
+    private final ClusterService clusterService;
+    private final Client client;
+
+    private final HealthService healthService;
+    private final Clock clock;
+
+    private final AtomicBoolean isHealthNode = new AtomicBoolean(false);
```

Comment:
I am wondering if we should use volatile here.... If I am not mistaken we do not perform any atomic operations on this flag. Are there any other reasons to prefer atomic over volatile?

Reply:
Oh, I think that this misunderstood my earlier comment about using an `AtomicBoolean` for protection from double running (in this comment: https://github.com/elastic/elasticsearch/pull/96772/#discussion_r1227308838)

We don't actually need the `AtomicBoolean` for the `isHealthNode` flag, as Mary said, `volatile` would be fine. What I meant about using an `AtomicBoolean` for protection is something more like this:

```java
private final AtomicBoolean currentlyRunning = new AtomicBoolean(false);
…
    public void triggered(SchedulerEngine.Event event) {
        if (event.getJobName().equals(HEALTH_PERIODIC_LOGGER_JOB_NAME)) {
            if (this.isHealthNode && this.currentlyRunning.compareAndSwap(false, true)) {
                this.callGetHealth();
            }
        }
    }
…

// And then much later in the listener for the results of the get-health call:

    protected final ActionListener<List<HealthIndicatorResult>> resultsListener = new ActionListener<List<HealthIndicatorResult>>() {
        @Override
        public void onResponse(List<HealthIndicatorResult> healthIndicatorResults) {
            …
            this.currentlyRunning.set(false);
        }

        @Override
        public void onFailure(Exception e) {
            …
            this.currentlyRunning.set(false);
        }
    };
```

This protects against us a case where the actual `callGetHealth()` call takes > 30 seconds. This atomic boolean would prevent us from ever invoking it more than once.

Hopefully that clarifies?